src/codegen/MHTranslator.cpp

```diff
@@ -1036,24 +1036,148 @@ void MHTranslator::transQuery(code::FunctionDecl* fun, std::shared_ptr<ir::Query
+    code::Expr* lhs = new code::Identifier(answervarname); // lhs duplicated
+    code::Expr* lhs2 = new code::Identifier(answervarname);
+    code::Expr* lhs3 = new code::Identifier(answervarname);
+    std::vector<code::Expr*> histArgs;
+    for (size_t i = 0; i < argName.size(); ++i) {
+      auto body = new code::CompoundStmt();
+      auto f_loop = new code::ForStmt(
+        new code::DeclStmt(new code::VarDecl(NULL,argName[i],INT_TYPE,new code::IntegerLiteral(0))),
+        new code::BinaryOperator(new code::Identifier(argName[i]),new code::IntegerLiteral(argDim[i]), code::OpKind::BO_LT),
+        new code::BinaryOperator(new code::Identifier(argName[i]), NULL, code::OpKind::UO_INC),
+        body);
+      context->addStmt(f_loop);
+      context = body;
+      // create multi-index
+      lhs = new code::ArraySubscriptExpr(lhs, new code::Identifier(argName[i]));
+      lhs2 = new code::ArraySubscriptExpr(lhs2, new code::Identifier(argName[i]));
+      lhs3 = new code::ArraySubscriptExpr(lhs3, new code::Identifier(argName[i]));
+      // create multi-argument list
+      histArgs.push_back(new code::Identifier(argName[i]));
+    }
+    context->addStmt(
+      new code::BinaryOperator(
+        lhs,
+        new code::NewExpr(initvalue),
+        code::OpKind::BO_ASSIGN));
+
+    // Declare Histogram Storage
+    code::Type histType =
+      code::Type(HISTOGRAM_CLASS_NAME,
+        std::vector<code::Type>( {
           (qr->getVar()->getTyp()->getTyp() == ir::IRConstant::BOOL ?
-            BOOL_TYPE : mapIRTypeToCodeType(qr->getVar()->getTyp())) })),
-      initvalue);
-  std::vector<code::Expr*> args;
-  args.push_back(transExpr(qr->getVar()));
-  args.push_back(new code::IntegerLiteral(1));
-  fun->addStmt(
-      code::CallExpr::createMethodCall(answervarname, HISTOGRAM_ADD_METHOD_NAME,
-                                       args));
-  // add print this result in print()
-  // :::=> answer_id.print("query expr");
-  corePrint->addStmt(
-      code::CallExpr::createMethodCall(answervarname,
-                                       HISTOGRAM_PRINT_METHOD_NAME,
-                                       std::vector<code::Expr*> { new code::StringLiteral(qr->str()) }));
+            BOOL_TYPE : mapIRTypeToCodeType(qr->getVar()->getTyp())) }), false, true);
+    code::VarDecl::createVarDecl(coreNs, answervarname,
+      code::Type(DYNAMICTABLE_CLASS_NAME,
+         std::vector<code::Type>{histType, code::Type(std::to_string(qr->getArgs().size()))}));
+
+    std::vector<code::Expr*> args;
+    args.push_back(transExpr(qr->getVar()));
+    args.push_back(new code::IntegerLiteral(1));
+    code::Stmt* evalSt = createPtrMethodCall(
+        lhs2, HISTOGRAM_ADD_METHOD_NAME, args);
+    code::Stmt* printSt = new code::CompoundStmt();
+    std::vector<code::Expr*> printExprVec = std::vector<code::Expr*>( {
+        new code::Identifier(buffername)});
+    std::string typeArgs = qr->getArg(0)->getTyp()->toString() + "[%d]";
```

Comment:
You should noticed that not all the instances necessarily have names in the form of "XXX[#digit#]".
I can write sth like 
BLOG:
distinct City A, B;
random A(City c) ~ ...
query A(c) for City c;

For a better approach, I have some suggestions:

<1>. quick and relatively nice solution:
Since we have restricted the for-loop-query to type with fixed amount of instances, the names of all the instances will be stored in "__vecstr_instance_XXXX" (see getInstanceStringArrayName(...) in translator.h).
So for the current moment, let's do sth like this:
BLOG: 
query A[c] for City c;
C++:
answer_0[c].print("A[c], c = " + __vecstr_instance_City[c] );

So in the terminal, the compiled target will print:
>>> A[c], c = A,
......
>>> A[c], c = B,
......

In the translator.cpp, let's create a small util function for generating the expression "A[c], c = " + __vecstr_instance_City[c]", which shouldn't be hard.
NOTE: do these things in the translator.cpp, and other translators and inherit these util functions and have concise code.

<2>. systematic solution
In the analyzer class, let's create some other functions/methods so that when creating the field "strExpr" in the class ir::Query, we can mark all those looping variables. Then in the translator, we replace those marked looping variables with with "__vecstr_instance_XXXX".

I think for the current moment, approach 1 should be sufficient and will lead to much shorter code.